ipq807x: add QNAP 301w

kirdesde opened the pull request:
QNAP 301w is a AX WIFI router with 4 1G and 2 10G ports.

Specifications:

    •     CPU: Qualcomm IPQ8072A Quad core Cortex-A53 2.2GHz
    •     RAM: 1024MB of DDR3
    •     Storage: 4GB eMMC (contains kernel and rootfs) / 8MB NOR
          (contains art and u-boot-env)
    •     Ethernet: 4x 1G RJ45 ports + 2 10G ports (Aquantia AQR113C)
    •     WLAN:
          2.4GHz: Qualcomm QCN5024 2x2 802.11b/g/n/ax 1174 Mbps PHY rate
          5GHz: Qualcomm QCN5054 4x4 802.11a/b/g/n/ac/ax 2402 PHY rate

    •     LEDs:
          7 x GPIO-controlled dual color LEDs + 2 GPIO-controlled single color LEDs

            • Buttons: 1x soft reset / 1x WPS
            • Power: 12V DC jack

        A poulated serial header is onboard.
        RX/TX is working, bootwait is active, secure boot is not enabled.

        SSH can be activated in the stock firmware, hold WPS button til the second beep
	(yes the router has a buzzer)
        SSH is available on port 22200, login with user admin and
	password "mac address of the router".

        Installation Instructions:

            • obtain serial access (https://openwrt.org/inbox/toh/qnap/301w#serial)
            • stop auto boot
            • setenv serverip 192.168.10.1
            • setenv ipaddr 192.168.10.10
            • tftpboot the initramfs image
              (openwrt-ipq807x-generic-qnap_301w-initramfs-fit-uImage.itb)
            • bootm

            • make sure that current_entry is set to "0":
              "fw_printenv -n current_entry" should be print "0". If not,
	       do "fw_setenv current_entry 0"
            • copy openwrt-ipq807x-generic-qnap_301w-squashfs-sysupgrade.bin
	          to the device to /tmp folder
            • sysupgrade -n /tmp/openwrt-ipq807x-generic-qnap_301w-squashfs-sysupgrade.bin
              this flashes openwrt to the first kernel and rootfs partition (mmcblk0p1 / mmcblk0p4)
            • reboot

        Note: this leaves the second kernel / rootfs parition untouched. So if you want
              to go back to stock, stop u-boot autoboot, "setenv current_entry 1" ,
	       "saveenv", "bootipq".
              Stock firmware should start from the second partition.
	      Then do a firmwareupgrade in the stock gui, that should overwrite the openwrt
	      in the first partitions

        Make 10G Aquantia phy's work:
              The aquantia phy's need a firmware to work. This can either be loaded
	      in linux with a userspace tool or in u-boot.

              I was not successfull to load the firmware in linux (aq-fw-download) but luckily there is
	      aq_load_fw available in u-boot. But first the right firmware needs to write
	      to the 0:ETHPHYFW mtd partition (it is empty on my device)

              Grab the ethphy firmware image from:
	      https://github.com/kirdesde/nbg7815_gpl/blob/master/target/linux/ipq/ipq807x_64/prebuilt_images/AQR_ethphyfw.mbn
	      and scp that to openwrt.

              Check the 0:ETHPHYFW partition number:
              cat /proc/mtd|grep "0:ETHPHYFW", should be mtd10.

              Backup the 0:ETHPHYFW partition:
              dd if=/dev/mtd10 of=/tmp/ethphyfw.backup, scp ethphyfw.backup to a save place.

              Write the new firmware image to the 0:ETHPHYFW partition:
              "mtd erase /dev/mtd10", "mtd -n write AQR_ethphyfw.mbn /dev/mtd10".

              Reboot to u-boot.

              Check if aq_load_fw is working:

              "aq_load_fw 0", that checks the firmware and if successfull,
	      loads iram and dram to one of the aquantia phy's.

              If that worked, add the aq_load_fw to the bootcmd:
              setenv "'aq_load_fw 0 && aq_load_fw 8 && bootipq'"

              "saveenv"

              "reset"

              Board reboots and the firmware load to both phy's should start and
	      then openwrt boots.

              Check if the 10G ports work.

              Note: lan port labeled "10G-2" is configured as WAN port as per default.
	      All other port are in the br-lan. This can be changed in the network config.

Signed-off-by: Dirk Buchwalder <buchwalder@posteo.de>

Flole998 commented:
The qnap flash stuff looks pretty similar to the netgear stuff that has been proposed in https://github.com/openwrt/openwrt/pull/3469 and merged in 2cb24b3f3cd89692f3c0bd137f3f560ada359bfa. Would it work to use the netgear stuff there? I don't really get why @chunkeey renamed it from mmc_do_flash as now we would be using netgear* named functions for qnap devices. Or we need to have the same code there twice. That can't be intended....

kirdesde commented:
> The qnap flash stuff looks pretty similar to the netgear stuff that has been proposed in [openwrt#3469](https://github.com/openwrt/openwrt/pull/3469) and merged in [2cb24b3](https://github.com/robimarko/openwrt/commit/2cb24b3f3cd89692f3c0bd137f3f560ada359bfa). Would it work to use the netgear stuff there? I don't really get why @chunkeey renamed it from mmc_do_flash as now we would be using netgear* named functions for qnap devices. Or we need to have the same code there twice. That can't be intended....

Hi, QNAP uses the qcom style kernel partition naming so we would at least change this:

`[ -z "$kernel" ] && kernel=$(find_mmc_part "kernel")`

to

`[ -z "$kernel" ] && kernel=$(find_mmc_part "kernel" | "0:HLOS")`

Kernel partitions QNAP:
`1	0x00000022	0x00008021	"0:HLOS"
	attrs:	0x0000000000000000
	type:	b51f2982-3ebe-46de-8721-ee641e1f9997
	guid:	1a373c3c-588e-eff4-5182-68ad83e0e62e
  2	0x00008022	0x00010021	"0:HLOS_1"
	attrs:	0x0000000000000000
	type:	a71da577-7f81-4626-b4a2-e377f9174525
	guid:	7d35f31d-0980-2220-edfd-6228595b79d2
  3	0x00010022	0x00018021	"0:HLOS_2"
	attrs:	0x0000000000000000
	type:	3e972198-336f-11ea-baa1-52540063a6b7
	guid:	964e7e67-6666-7303-312a-291b757640e8`

But you are right, it's a pitty that the upgrade script is now netgear related and ipq40xx target specific. A more generic approach would be better (like your proposal)
And due to the netgear.sh is target specific and ipq807x support isn't official at the moment, i would rather stick to my qnap.sh for the moment.
Even if we are using the same code, we need to maintain both scripts separately due to different targets.

robimarko commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -50,6 +50,7 @@ ALLWIFIBOARDS:= \
 	nec_wg2600hp3 \
 	plasmacloud_pa1200 \
 	plasmacloud_pa2200 \
+  qnap_301w \
```
Broken indent

kirdesde replied:
fixed

robimarko commented on target/linux/ipq807x/base-files/lib/upgrade/qnap.sh:
```diff
@@ -0,0 +1,81 @@
+	case "$board" in
+	qnap,301w)
+			kernel="/dev/mmcblk0p1"
+			rootfs="/dev/mmcblk0p4"
```
Can these be looked up by name instead of hardcoding?

kirdesde replied:
I could use the lately committed "find_mmc_part" function, but that would need a rebase of your branch on openwrt master.
Or I can implement something similar in the qnap.sh and can switch later to the find_mmc_part function if it's in your branch.

kirdesde replied:
I added the "find_mmc_part" function from openwrt master to the qnap.sh and switched to partition name.
If the generic "find_mmc_part" gets merged to your branch in the future, I'll remove the "find_mmc_part" from the qnap.sh.

robimarko replied:
I have rebased the branch, so please use the upstream function and remove the hardcoded path as OpenWrt has an issue with those and if anything changes in the block device numbering then it simply breaks updating.

kirdesde replied:
done. rebased, using upstream function and removed hardcoded path (using part name instead)

robimarko replied:
Is this script really special to QNAP or is it mostly copy/paste?

kirdesde replied:
It's mostly c&p from the zyxel.sh from the ipq806x target.

I think this could be generic, and we only need to pass kernel part name and rootfs name to the script.

And the loopdev offset check here:

```
local offset=$(tar xf $tar_file ${board_dir}/root -O | wc -c)
	[ $offset -lt **131072** ] 
```
The bold value needs to align to the blocksize in the image definition (otherwise openwrt doesn't recognize the ext4 overlay loopdev). But that's all. So we could maybe define that 3 as variables in the platform.sh and pass that to the script.

robimarko replied:
That would be great to avoid duplicating the code

kirdesde replied:
Hi Robert,

I've pushed a few changes.

1. a mmc.sh as a generic mmc upgrade script
2. in platform.sh I've added a variable "kernelname" and "rootfsname" they will later be used in in the function "mmc_do_upgrade" (mmc.sh)
3. this means, the mmc_do_uggrade may also used for other devices, just define kernelname and rootfsname in platform.sh and call mmc_do_uggrade.

I hopes this is fine.

> > I am fine with having QNAP specific update script.
> 
> And then have basically the same script for netgear aswell, just with a different partition name in case that's needed? Sounds a little unusual to me, just wanted to double-check if I understood that correctly.

Hi, now we have generic mmc.sh, I hoped that is what you're looking for ;-)

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+	/*
+	 *remove qpic pins due to device is using all of the pins for leds-gpio
+	 */
+	qpic_pins: qpic-pins {
```
Can you try if adding status = "disabled" works instead?
And if not, then delete the whole node?

kirdesde replied:
yes I can check that.

I already tried deleting that node, but that leads to an error from the dtc due to qpic-nand is using the qpic-pins. Means deleting the entire nand node would also necessary.

I'll try disabling the qpic-pins node first.

kirdesde replied:
disabling worked. Led's still working.

robimarko replied:
Great, please use that.

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+&blsp1_spi1 { /* BLSP1 QUP1 */
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	cs-select = <0>;
```
This doesn't exist, use cs-gpios instead

kirdesde replied:
sorry, can you give me a hint? I can't find cs-gpios in the ipq8074.dtsi either.
I would just remove that property.

robimarko replied:
You cant remove it, it's required.
You won't find it in the DTSI as its device-specific, here is the binding:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/devicetree/bindings/spi/spi-controller.yaml?h=v5.15.1#n28

In your case `cs-gpios = <0>;` should be what you are looking for as that just means to use the HW CS

kirdesde replied:
thanks for the explanation, added.

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	cs-select = <0>;
+	status = "ok";
```
"ok" is per DTC spec, use "okay"

kirdesde replied:
done

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+	cs-select = <0>;
+	status = "ok";
+
+	m25p80@0 {
```
flash@0

kirdesde replied:
done

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+		#address-cells = <1>;
+		#size-cells = <1>;
+		reg = <0>;
+		compatible = "n25q128a11";
```
Use "jedec,spi-nor"

kirdesde replied:
done

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+		#size-cells = <1>;
+		reg = <0>;
+		compatible = "n25q128a11";
+		linux,modalias = "m25p80", "n25q128a11";
```
This is again just QSDK crap, remove it

kirdesde replied:
done

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+		compatible = "n25q128a11";
+		linux,modalias = "m25p80", "n25q128a11";
+		spi-max-frequency = <50000000>;
+		use-default-sizes;
```
More QSDK crap, this doesnt exist

kirdesde replied:
done

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+		linux,modalias = "m25p80", "n25q128a11";
+		spi-max-frequency = <50000000>;
+		use-default-sizes;
+		m25p,fast-read;
```
Are you sure that the SPI-NOR supports it?
What's the model inside?

kirdesde replied:
No I'm not sure...model is Winbond W25Q64DW 
I'll remove this.

robimarko replied:
It probably does, but since its not actually used as the rootfs storage I would remove it anyway.

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+	ethernet-phy@8 {
+		compatible ="ethernet-phy-ieee802.3-c45";
+		reg = <8>;
+        reset-gpios = <&tlmm 44 GPIO_ACTIVE_LOW>;
```
Broken indent

kirdesde replied:
done

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+    ethernet-phy@0 {
+		compatible ="ethernet-phy-ieee802.3-c45";
+		reg = <0>;
+        reset-gpios = <&tlmm 59 GPIO_ACTIVE_LOW>;
```
Broken indent

robimarko replied:
Also, can you sort them by reg property

kirdesde replied:
both done

kirdesde replied:
Only one thing that is not working is the second 10G phy and and don't know why. I'm pretty sure this was working a few days ago but now there is only a physical link and link activity (according to the phy leds) but there is no ip traffic.

I already tried switching the gpio-reset gpio's, removing the reset-gpios...but it's still not working.
Maybe you have an idea.

The first 10G works like a charm.

robimarko replied:
I would bet that it's due to the reset GPIO not actually being toggled fast enough.
You can try adding a reset-post-delay-us property so it waits X amount of time before trying to probe after reset is de-asserted.
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/Documentation/devicetree/bindings/net/mdio.yaml?h=v5.15.1#n42

I had the same issue on AX9000 where the QCA8081 wasn't replying fast enough for auto-ID, but I then set the PHY ID directly in the compatible and it works that way.
https://github.com/robimarko/openwrt/blob/IPQ807x-5.10-backports/target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-ax9000.dts#L227

kirdesde replied:
This is really strange...the second 10G port works only as wan (uplink) port.

I can't get them to work as LAN-Port.

I tried both of your solutions, without any luck.

So I changed the lan/wan port definition and changed the commit messages regarding the default network configuration.

robimarko replied:
I think I know whats the issue, it's the crap QCA switch driver.
As they are not using DSA they are using the following DTS properties:
```
switch_lan_bmp = <0x3e>; /* lan port bitmap */
switch_wan_bmp = <0xc0>; /* wan port bitmap */
```

According to these then SSDK will configure the forwarding register and allow/disallow forwarding between certain ports.
Usually the WAN port can only forward to the CPU port, hence why you cant use it as LAN

kirdesde replied:
Thanks for the explanation, so this means is is hardcoded in the dts.

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+};
+
+&pcie0 {
+	status = "disabled";
```
This is disabled by default

kirdesde replied:
removed

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+};
+
+&pcie1 {
+	status = "disabled";
```
Also disabled by default

kirdesde replied:
removed

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+
+&wifi {
+	status = "okay";
+	qcom,board_id = <0xff>;
```
0xff is the default value, I highly doubt that they use that in stock FW.
Its meant to be the fallback

kirdesde replied:
according to the stock bootlog they are actually using the 0xff:

cnss: Target capability: chip_id: 0x0, chip_family: 0x0, board_id: 0xff, soc_id: 0xffffffff, fw_version: 0x220284a5, fw_build_timestamp: 2020-06-05 00:18, otp_version: 0x0 eeprom_caldata_read_timeout 0s
[   40.039493] cnss: Downloading BDF: bdwlan.bin, size: 64
[   40.057456] cnss: No board_id entry in device tree

robimarko replied:
Ok, this is the first one then.
Then leave the board_id setting and please add a comment that the stock FW uses it and that its intentional so it doesn't get removed by someone else later.

kirdesde replied:
done

robimarko commented on target/linux/ipq807x/image/generic.mk:
```diff
@@ -36,6 +42,19 @@ define Device/edimax_cax1800
+	DEVICE_MODEL := 301w
+	DEVICE_DTS_CONFIG := config@hk01
+	KERNEL_SIZE := 16384k
+	BLOCKSIZE := 64k
```
Whats the eMMC inside as this looks real low for a modern eMMC

kirdesde replied:
It's a MTFC4GACAJCN-1M

kirdesde replied:
Hmm, I haven't found anything blocksize related in the public available datasheet, but maybe I'm overlooking something

kirdesde replied:
"mmc extcsd read" list this:

High-capacity erase unit size [HC_ERASE_GRP_SIZE: 0x01]
 i.e. 512 KiB
 
 I guess this can be used as optimal blocksize?

robimarko replied:
Since it's eMMC, it probably doesn't care about the FS block size as its FW will handle that internally anyway, but 64k is too small.
I would go for 128k at least, what does the stock FW use?

kirdesde replied:
Good question, how can I figure this out?

robimarko replied:
What's the format of the stock FW?

kirdesde replied:
GPL doesn't contain anything regarding image build.

So I can only reference the fwupdate.sh.

Stock firmware contains 3 squashfs images: kernel, rootfs and userfs

As for the kernel partition they delete with a blocksize of 512 and then cp the kernel partition:

```
dd if=/dev/zero bs=512 count=${part_block_num} of=${next_kernel_part} || echo_dbg_and_exit "earse ${next_kernel_part} fail" ${ERRCODE_WRITE_KERNEL}
    echo_dbg "write kernel to device node ${next_kernel_part} ....."
    dd if=${FW_KERNEL} of=${next_kernel_part} || echo_dbg_and_exit "write kernel to ${next_kernel_part} fail" ${ERRCODE_WRITE_KERNEL}
```
    
Regarding the rootfs partition:
Just a mke2fs with ext4 (FSTYPE_ROOTFS), nothing blocksize specific as far as I can see.

```

#if the partition is empty and not ext4 file system, format it using mke2fs
    if [ x"${next_rootfs_part}" != x"" ] && [ -e ${next_rootfs_part} ] && ! blkid ${next_rootfs_part} | grep -q ${FSTYPE_ROOTFS} ; then
        echo_dbg "the filesystem of ${dev_rootfs_data} is not ${FSTYPE_ROOTFS}, format it !!!"
        #if it partition is other partition, it should use yes to make sure to format it
        yes | mkfs.${FSTYPE_ROOTFS} ${next_rootfs_part}
    fi

    #because the initial file system image maybe not fit partition size in factory, I resize to max in fw upgrade if need
    resize2fs -fp ${next_rootfs_part}

    mount ${next_rootfs_part} ${ROOTFS_MNTPNT} || echo_dbg_and_exit "mount ${next_rootfs_part} ${ROOTFS_MNTPNT} fail" ${ERRCODE_MOUNT_ROOTFS_PART}
    rm -rf ${ROOTFS_MNTPNT}/*
    cp ${FW_ROOTFS} ${FW_USR} ${ROOTFS_MNTPNT} || echo_dbg_and_exit "write kernel to ${next_rootfs_part} fail" ${ERRCODE_WRITE_ROOTFS}
    umount ${ROOTFS_MNTPNT}
}
```
Then they mount that partition and cp rootfs.squashfs and usr.squashfs to the mountpoint.

`resize2fs -fp ${next_rootfs_part}

    mount ${next_rootfs_part} ${ROOTFS_MNTPNT} || echo_dbg_and_exit "mount ${next_rootfs_part} ${ROOTFS_MNTPNT} fail" ${ERRCODE_MOUNT_ROOTFS_PART}
    rm -rf ${ROOTFS_MNTPNT}/*
    cp ${FW_ROOTFS} ${FW_USR} ${ROOTFS_MNTPNT} || echo_dbg_and_exit "write kernel to ${next_rootfs_part} fail" ${ERRCODE_WRITE_ROOTFS}
    umount ${ROOTFS_MNTPNT}`

kirdesde replied:
I've increased the blocksize to 128k

kirdesde replied:
I've increased the blocksize to 512k

robimarko commented on target/linux/ipq807x/image/generic.mk:
```diff
@@ -23,6 +23,12 @@ define Device/UbiFit
 	IMAGE/nand-sysupgrade.bin := sysupgrade-tar | append-metadata
 endef
 
+define Device/QnapImage
```
Is there a generic recipe for these, they don't really look special to me at all

kirdesde replied:
Not as I know it.  I find this incl. the padding only device specific in other targets.
Should I remove the image recipe and move the image generation to the device/qnap_301w?

kirdesde replied:
I've removed the QnapImage recipe and added the image definition to the device itself. I hope it's allright.

robimarko replied:
That is fine as there is only one device using it, if there are multiple then it makes sense to just reference the recipe

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+	};
+
+	dp6: dp6 {
+			device_type = "network";
```
Broken indent

kirdesde replied:
done

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+};
+
+&qpic_nand {
+	status = "disabled";
```
This should be disabled by default

kirdesde replied:
removed

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+	status = "ok";
+};
+
+
```
Extra newline

kirdesde replied:
removed

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+};
+
+&ssphy_0 {
+	status = "ok";
```
More "ok" instead of "okay"-s

kirdesde replied:
done

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+
+	
+};
+
```
Extra newline

kirdesde replied:
removed

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-301w.dts:
```diff
@@ -0,0 +1,836 @@
+
+	aliases {
+		serial0 = &blsp1_uart5;
+		sdhc1 = &sdhc_1;
```
Is this even needed?

kirdesde replied:
removed

Flole998 commented:
> > The qnap flash stuff looks pretty similar to the netgear stuff that has been proposed in [openwrt#3469](https://github.com/openwrt/openwrt/pull/3469) and merged in [2cb24b3](https://github.com/robimarko/openwrt/commit/2cb24b3f3cd89692f3c0bd137f3f560ada359bfa). Would it work to use the netgear stuff there? I don't really get why @chunkeey renamed it from mmc_do_flash as now we would be using netgear* named functions for qnap devices. Or we need to have the same code there twice. That can't be intended....
> 
> Hi, QNAP uses the qcom style kernel partition naming so we would at least change this:
> 
> `[ -z "$kernel" ] && kernel=$(find_mmc_part "kernel")`
> 
> to
> 
> `[ -z "$kernel" ] && kernel=$(find_mmc_part "kernel" | "0:HLOS")`
> 
> Kernel partitions QNAP:
> 
> ```
>   1	0x00000022	0x00008021	"0:HLOS"
> 	attrs:	0x0000000000000000
> 	type:	b51f2982-3ebe-46de-8721-ee641e1f9997
> 	guid:	1a373c3c-588e-eff4-5182-68ad83e0e62e
>   2	0x00008022	0x00010021	"0:HLOS_1"
> 	attrs:	0x0000000000000000
> 	type:	a71da577-7f81-4626-b4a2-e377f9174525
> 	guid:	7d35f31d-0980-2220-edfd-6228595b79d2
>   3	0x00010022	0x00018021	"0:HLOS_2"
> 	attrs:	0x0000000000000000
> 	type:	3e972198-336f-11ea-baa1-52540063a6b7
> 	guid:	964e7e67-6666-7303-312a-291b757640e8
> ```
> 
> But you are right, it's a pitty that the upgrade script is now netgear related and ipq40xx target specific. A more generic approach would be better (like your proposal) And due to the netgear.sh is target specific and ipq807x support isn't official at the moment, i would rather stick to my qnap.sh for the moment. Even if we are using the same code, we need to maintain both scripts separately due to different targets.

As I'm working on some IPQ807x Netgear devices that probably need the same or a similar way of upgrading it would be great if I could use your work. Maybe it's a good idea to make the name for the partitions configurable with a variable.

kirdesde commented:
> > > The qnap flash stuff looks pretty similar to the netgear stuff that has been proposed in [openwrt#3469](https://github.com/openwrt/openwrt/pull/3469) and merged in [2cb24b3](https://github.com/robimarko/openwrt/commit/2cb24b3f3cd89692f3c0bd137f3f560ada359bfa). Would it work to use the netgear stuff there? I don't really get why @chunkeey renamed it from mmc_do_flash as now we would be using netgear* named functions for qnap devices. Or we need to have the same code there twice. That can't be intended....
> > 
> > 
> > Hi, QNAP uses the qcom style kernel partition naming so we would at least change this:
> > `[ -z "$kernel" ] && kernel=$(find_mmc_part "kernel")`
> > to
> > `[ -z "$kernel" ] && kernel=$(find_mmc_part "kernel" | "0:HLOS")`
> > Kernel partitions QNAP:
> > ```
> >   1	0x00000022	0x00008021	"0:HLOS"
> > 	attrs:	0x0000000000000000
> > 	type:	b51f2982-3ebe-46de-8721-ee641e1f9997
> > 	guid:	1a373c3c-588e-eff4-5182-68ad83e0e62e
> >   2	0x00008022	0x00010021	"0:HLOS_1"
> > 	attrs:	0x0000000000000000
> > 	type:	a71da577-7f81-4626-b4a2-e377f9174525
> > 	guid:	7d35f31d-0980-2220-edfd-6228595b79d2
> >   3	0x00010022	0x00018021	"0:HLOS_2"
> > 	attrs:	0x0000000000000000
> > 	type:	3e972198-336f-11ea-baa1-52540063a6b7
> > 	guid:	964e7e67-6666-7303-312a-291b757640e8
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But you are right, it's a pitty that the upgrade script is now netgear related and ipq40xx target specific. A more generic approach would be better (like your proposal) And due to the netgear.sh is target specific and ipq807x support isn't official at the moment, i would rather stick to my qnap.sh for the moment. Even if we are using the same code, we need to maintain both scripts separately due to different targets.
> 
> As I'm working on some IPQ807x Netgear devices that probably need the same or a similar way of upgrading it would be great if I could use your work. Maybe it's a good idea to make the name for the partitions configurable with a variable.

Hi, of course you can use my work ;-)

Like I use your find_mmc_part function, I added that to the qnap.sh and added a case - esac statement to make sure, the upgrade is running on the right device.

case "$board" in
	qnap,301w)

Sure, we can also use variables for the part names.

Flole998 commented:
That find_mmc_part function wasn't written by me ;)

If we want to use it on qnap and netgear devices then it should be more like a mmc_do_upgrade/mmc_do_flash in a mmc.sh file as I don't think it is a good idea to have stuff used by netgear in a file for qnap.

robimarko commented:
I have rebased my branch so that you can use the MMC generic functions, so please rebase and use those.
I am fine with having QNAP specific update script.

Flole998 commented:
> I am fine with having QNAP specific update script.

And then have basically the same script for netgear aswell, just with a different partition name in case that's needed? Sounds a little unusual to me, just wanted to double-check if I understood that correctly.

Flole998 commented:
Also is IPQ807x-5.10-backports now the "new" branch which should be used for further work?

robimarko commented:
Oh, if it's the same code then please don't duplicate it for the sake of naming, rather have a commit that makes it generic and just use it.
Yeah, the backports branch is the current one to use

kirdesde commented:
> > I am fine with having QNAP specific update script.
> 
> And then have basically the same script for netgear aswell, just with a different partition name in case that's needed? Sounds a little unusual to me, just wanted to double-check if I understood that correctly.


Hi, now we have generic mmc.sh, I hoped that is what you're looking for ;-)

Flole998 commented:
Thanks! That should make my work a lot easier once I finally find some time to finish it.

kirdesde commented on target/linux/ipq807x/image/generic.mk:
```diff
@@ -36,6 +36,21 @@ define Device/edimax_cax1800
+	BLOCKSIZE := 512k
+	SOC := ipq8072
+	IMAGES += factory.bin sysupgrade.bin
+	IMAGE/factory.bin := append-rootfs | pad-rootfs | pad-to 64k
```
Yes, "pad-to 64k" is the only way I'v got the loopdev overlay working. With any other padding size, openwrt doesn't recognize the overlay and falls back to /tmp (no matter what blocksize the image is using)

kirdesde commented on target/linux/ipq807x/image/generic.mk:
```diff
@@ -36,6 +36,21 @@ define Device/edimax_cax1800
+	IMAGES += factory.bin sysupgrade.bin
+	IMAGE/factory.bin := append-rootfs | pad-rootfs | pad-to 64k
+	IMAGE/sysupgrade.bin/squashfs := append-rootfs | pad-to 64k | sysupgrade-tar rootfs=$$$$@ | append-metadata
+	DEVICE_PACKAGES := ipq-wifi-qnap_301w e2fsprogs kmod-fs-ext4 losetup uboot-envtools kmod-loop
```
okay, then I'll remove that.

Losetup is needed to created the actual loop dev after flashing:

`local loopdev="$(losetup -f)"
	losetup -o $offset $loopdev $rootfs`

kirdesde replied:
kmod-loop removed

robimarko replied:
Ok, thats fine then.

robimarko commented:
Can you just let me know once you have pushed everything you wanted.
I think this is ready

kirdesde commented:
> Can you just let me know once you have pushed everything you wanted. I think this is ready

It's ready, I have nothing more to push.
